game-core/src/main/java/games/strategy/engine/data/GameMap.java

```diff
@@ -223,6 +224,17 @@ public Territory getTerritory(final String s) {
     return getNeighbors(frontier, searched, distance, null);
   }
 
+  Set<Territory> getNeighborsValidatingCanals(final Territory territory, final Predicate<Territory> neighborFilter,
+      final Collection<Unit> units, final PlayerID player) {
+    final Set<Territory> neighbors = getNeighbors(territory, neighborFilter);
+    if (player != null) {
+      return neighbors.parallelStream()
+          .filter(t -> MoveValidator.canAnyUnitsPassCanal(territory, t, units, player, getData()))
+          .collect(Collectors.toSet());
+    }
+    return neighbors;
+  }
+
```

Comment:
Parallel stream did seem to make some performance improvement here. This most likely should be used in a lot of other places especially in the AI but gotta start somewhere.